modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ExecutionServiceImpl.java

```diff
@@ -842,15 +843,29 @@ private boolean single(SqlNode sqlNode) {
     private void onMessage(UUID nodeId, QueryStartRequest msg) {
         assert nodeId != null && msg != null;
 
-        PlanningContext pctx = createContext(Contexts.empty(), msg.topologyVersion(), nodeId, msg.schema(), msg.root(), msg.parameters());
+        try {
+            PlanningContext pctx = createContext(Contexts.empty(), msg.topologyVersion(), nodeId, msg.schema(), msg.root(), msg.parameters());
 
-        List<QueryPlan> qryPlans = queryPlanCache().queryPlan(pctx, new CacheKey(pctx.schemaName(), pctx.query()), this::prepareFragment);
+            List<QueryPlan> qryPlans = queryPlanCache().queryPlan(
+                pctx,
+                new CacheKey(pctx.schemaName(), pctx.query()),
+                this::prepareFragment
+            );
 
-        assert qryPlans.size() == 1 && qryPlans.get(0).type() == QueryPlan.Type.FRAGMENT;
+            assert qryPlans.size() == 1 && qryPlans.get(0).type() == QueryPlan.Type.FRAGMENT;
 
-        FragmentPlan plan = (FragmentPlan)qryPlans.get(0);
+            FragmentPlan plan = (FragmentPlan)qryPlans.get(0);
 
-        executeFragment(msg.queryId(), plan, pctx, msg.fragmentDescription());
+            executeFragment(msg.queryId(), plan, pctx, msg.fragmentDescription());
+        }
+        catch (Throwable ex) {
+            try {
+                exchangeSvc.sendError(nodeId, msg.queryId(), msg.fragmentId(), ex);
```

Comment:
It would be ok to skip sending an answer or log an error if ExecService is about to stop.
WDYT?

Reply:
Error (e.g. OOME) should be rethrown to allow FH to kill current node.

Reply:
> It would be ok to skip sending an answer or log an error if ExecService is about to stop.
WDYT?

yes, it absolutely ok. All awaiting queries should be cancelled by originator node when NODE_LEFT event will arise. I'd rather log it with debug or trace level

> Error (e.g. OOME) should be rethrown to allow FH to kill current node.

sounds legit